src/content/developers/docs/smart-contracts/testing/index.md

```diff
@@ -1,37 +1,217 @@
+
+Manual testing of smart contracts requires considerable skill and a considerable investment of time, money, and effort. Moreover, manual testing can sometimes be susceptible to the problems of human error.  
+
+However, applying manual testing to smart contracts can also be beneficial. Code audits harness human intelligence to find defects in contract code that might go undetected during automated testing.
+
+Manual-testing your smart contracts can also reveal vulnerabilities that exist outside the code, but can still affect it. For example, a smart contract audit can discover vulnerabilities arising from flawed interaction with off-chain components. 
+
+## Why is it important to test smart contracts? {#benefits-of-smart-contract-testing}
+
+Testing smart contracts is important for the following reasons:
+
+### 1. Smart contracts are high-value applications {#smart-contracts-are-high-value-applications}
+
+Smart contracts often deal with high-value financial assets, especially in industries like [decentralized finance (DeFi)](/defi/), and valuable items, such as [non-fungible tokens (NFTs)](/nft/). As such, minor vulnerabilities in smart contracts can and often lead to massive, irrecoverable losses for users. Comprehensive testing can, however, expose errors in smart contract code and reduce security risks before deployment. 
+
+### 2. Smart contracts are immutable {#smart-contracts-are-immutable}
+
+Smart contracts deployed in the [Ethereum Virtual Machine (EVM)](/developers/docs/evm/) are immutable by default. While traditional developers may be used to fixing software bugs after launching, Ethereum development leaves little room for patching security flaws once a smart contract is live on the blockchain. 
+
+While upgradeability mechanisms for smart contracts, such as proxy patterns, these can be difficult to implement. Besides reducing immutability and introducing complexity, upgrades often demand complex governance processes. 
+
+For the most part, upgrades should be considered a last resort and avoided unless necessary.  Detecting potential vulnerabilities and flaws in your smart contract during the pre-launch phase reduces the need for a logic upgrade. 
+
+## Automated testing for smart contracts {#automated-testing-for-smart-contracts}
+
+### 1. Functional testing {#functional-testing}
+
+Functional testing verifies the functionality of a smart contract and provides assurance that each function in the code works as expected. Functional testing requires understanding how your smart contract should behave in certain conditions. Then you can test each function by running computations with selected values and comparing the returned output with the expected output.   
+
+Functional testing covers three methods: **unit testing**, **integration testing**, and **system testing**.
+
+- **Unit testing**
```

Comment:
```suggestion
#### Unit testing
```

Let's use H4s here - using bullets for this header sections looks weird to me:

<img width="623" alt="Screen Shot 2022-06-14 at 9 26 55 AM" src="https://user-images.githubusercontent.com/8097623/173588518-04cf12da-de8a-4abc-965b-8ea73bdccc93.png">